ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java

```diff
@@ -0,0 +1,242 @@
+                              return eip1559.computeBaseFee(
+                                  nextBlockNumber,
+                                  explicitlyRequestedBaseFees.get(
+                                      explicitlyRequestedBaseFees.size() - 1),
+                                  lastBlockHeader.getGasUsed(),
+                                  eip1559.targetGasUsed(lastBlockHeader));
+                            })
+                        .orElse(0L));
+
+    final List<Double> gasUsedRatios =
+        blockHeaders.stream()
+            .map(blockHeader -> blockHeader.getGasUsed() / (double) blockHeader.getGasLimit())
+            .collect(toUnmodifiableList());
+
+    final Optional<List<List<Long>>> maybeRewards =
+        maybeRewardPercentiles.map(
+            rewardPercentiles ->
+                LongStream.range(oldestBlock, oldestBlock + blockCount)
+                    .mapToObj(blockNumber -> blockchain.getBlockByNumber(blockNumber).get())
+                    .map(block -> computeRewards(rewardPercentiles, block))
+                    .collect(toUnmodifiableList()));
+
+    return new JsonRpcSuccessResponse(
+        requestId,
+        new FeeHistory(
+            oldestBlock,
+            Stream.concat(explicitlyRequestedBaseFees.stream(), Stream.of(nextBaseFee))
+                .collect(toUnmodifiableList()),
+            gasUsedRatios,
+            maybeRewards));
+  }
+
+  private List<Long> computeRewards(
+      final List<Double> rewardPercentiles, final org.hyperledger.besu.ethereum.core.Block block) {
+    final List<Transaction> transactions = block.getBody().getTransactions();
+    if (transactions.isEmpty()) {
+      // all 0's for empty block
+      return LongStream.generate(() -> 0)
+          .limit(rewardPercentiles.size())
+          .boxed()
+          .collect(toUnmodifiableList());
+    }
+
+    final Optional<Long> baseFee = block.getHeader().getBaseFee();
+    final List<Transaction> transactionsAscendingEffectiveGasFee =
+        transactions.stream()
+            .sorted(
+                Comparator.comparing(
+                    transaction -> transaction.getEffectivePriorityFeePerGas(baseFee)))
+            .collect(toUnmodifiableList());
+
+    final ArrayList<Long> rewards = new ArrayList<>();
+    int rewardIndex = 0;
+    long gasUsed = 0;
+    for (final Transaction transaction : transactionsAscendingEffectiveGasFee) {
+
+      gasUsed +=
+          blockchainQueries
+              .transactionReceiptByTransactionHash(transaction.getHash())
+              .get()
+              .getGasUsed();
+
+      while (rewardIndex < rewardPercentiles.size()
+          && 100.0 * gasUsed / block.getHeader().getGasUsed()
+              >= rewardPercentiles.get(rewardIndex)) {
+        rewards.add(transaction.getEffectivePriorityFeePerGas(baseFee));
+        rewardIndex++;
+      }
+    }
+    return rewards;
+  }
+
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public static class FeeHistory {
```

Comment:
NIT: For consistency with the current code and specification, I would say that it's better to rename the class to FeeHistoryResult and to use the Jackson and Immutable annotations

Spec : https://github.com/ethereum/eth1.0-specs/pull/212/files#diff-c56024dc5c5982b3df8a0e683e2e340ff092b384030ce455e4a4b45239a43b81R235
Location in the code : https://github.com/hyperledger/besu/tree/e230a445bb457bbac27bf4c384da449ccbd3f043/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results
Sample of result class : https://github.com/hyperledger/besu/tree/e230a445bb457bbac27bf4c384da449ccbd3f043/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results